rclcpp/include/rclcpp/qos_event.hpp

```diff
@@ -0,0 +1,164 @@
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__QOS_EVENT_HPP_
+#define RCLCPP__QOS_EVENT_HPP_
+
+#include <functional>
+
+#include "rcl/error_handling.h"
+
+#include "rcutils/logging_macros.h"
+
+#include "rclcpp/waitable.hpp"
+#include "rclcpp/exceptions.hpp"
+#include "rclcpp/function_traits.hpp"
+
+
+namespace rclcpp
+{
+
+struct QOSDeadlineEventInfo
+{
+  enum DeadlineEventType
+  {
+    DEADLINE_MISSED
+  };
+
+  DeadlineEventType event_type;
+  void * other_info;
+};
+
+struct QOSLivelinessEventInfo
+{
+  enum LivelinessEventType
+  {
+    LIVELINESS_CHANGED
+  };
+
+  LivelinessEventType event_type;
+  void * other_info;
+};
+
+struct QOSLifespanEventInfo
+{
+  enum LifespanEventType
+  {
+    LIFESPAN_EXPIRED
+  };
+
+  LifespanEventType event_type;
+  void * other_info;
+};
+
+
+using QOSDeadlineEventCallbackType = std::function<void (QOSDeadlineEventInfo &)>;
+using QOSLivelinessEventCallbackType = std::function<void (QOSLivelinessEventInfo &)>;
+using QOSLifespanEventCallbackType = std::function<void (QOSLifespanEventInfo &)>;
+
+
+// struct QOSDeadlineEventConfig {
+//   using CallbackT = QOSDeadlineEventCallbackType;
+//   using EventInfoT = QOSDeadlineEventInfo;
+// };
+
+// struct QOSLivelinessEventConfig {
+//   using CallbackT = QOSLivelinessEventCallbackType;
+//   using EventInfoT = QOSLivelinessEventInfo;
+// };
+
+// struct QOSLifespanEventConfig {
+//   using CallbackT = QOSLifespanEventCallbackType;
+//   using EventInfoT = QOSLifespanEventInfo;
+// };
+
+
+class QOSEventBase : public Waitable
+{
+public:
+  /// Get the number of ready events
+  size_t
+  get_number_of_ready_events() override;
+
+  /// Add the Waitable to a wait set.
+  bool
+  add_to_wait_set(rcl_wait_set_t * wait_set) override;
+
+  /// Check if the Waitable is ready.
+  bool
+  is_ready(rcl_wait_set_t * wait_set) override;
+
+protected:
+  rcl_event_t event_handle_;
+  size_t wait_set_event_index_;
+};
+
+
+template<typename EventCallbackT>
+class QOSEvent : public QOSEventBase
```

Comment:
Kind of a nitpick, but I think we should come up with a better name than QOSEvent as the name implies it is an occurrence of an event whereas this appears to be a container for everything related to the event handler.

Reply:
As @ross-desmond suggested, I will rename this to QOSEventHandler.